kura/org.eclipse.kura.wire.component.provider/src/main/java/org/eclipse/kura/internal/wire/subscriber/CloudSubscriber.java

```diff
@@ -90,43 +144,63 @@
      */
     protected synchronized void activate(final ComponentContext componentContext,
             final Map<String, Object> properties) {
-        s_logger.debug(s_message.activatingCloudSubscriber());
+        logger.debug(wireMessages.activatingCloudSubscriber());
+        this.componentContext = componentContext;
         this.wireSupport = this.wireHelperService.newWireSupport(this);
         this.options = new CloudSubscriberOptions(properties);
-        this.topic = this.options.getSubscribingTopic();
+        this.appTopic = this.options.getSubscribingAppTopic();
+        this.deviceId = this.options.getSubscribingDeviceId();
+        initCloudServiceTracking();
+        logger.debug(wireMessages.activatingCloudSubscriberDone());
+    }
+
+    private void subscribeTopic() throws KuraException {
+        if (this.cloudService.isConnected() && this.cloudClient != null) {
+            this.cloudClient.subscribe(this.deviceId, this.appTopic, this.options.getSubscribingQos());
+        }
+    }
+
+    /**
+     * Service tracker to manage Cloud Services
+     */
+    private void initCloudServiceTracking() {
+        String selectedCloudServicePid = this.options.getCloudServicePid();
+        String filterString = String.format("(&(%s=%s)(kura.service.pid=%s))", Constants.OBJECTCLASS,
+                CloudService.class.getName(), selectedCloudServicePid);
+        Filter filter = null;
         try {
-            if (this.dataService.isConnected()) {
-                this.dataService.subscribe(this.topic, this.options.getSubscribingQos());
-            }
-        } catch (final KuraException e) {
-            s_logger.error(s_message.errorCreatingCloudClinet() + e);
+            filter = this.componentContext.getBundleContext().createFilter(filterString);
```

Comment:
extract bundle context in an instance or member variable to save processing `getBundleContext` on `ComponentContext` multiple times